iceoryx2-cal/src/dynamic_storage/posix_shared_memory.rs

```diff
@@ -186,75 +313,85 @@ impl<T: Send + Sync + Debug> DynamicStorageBuilder<T, Storage<T>> for Builder<T>
+{
+    fn has_ownership(mut self, value: bool) -> Self {
+        self.has_ownership = value;
+        self
     }
 
-    fn try_open(self) -> Result<Storage<T>, DynamicStorageOpenError> {
-        let msg = "Failed to open ";
+    fn initializer<F: FnMut(&mut T, &mut BumpAllocator) -> bool + 'builder>(
+        mut self,
+        value: F,
+    ) -> Self {
+        self.initializer = Initializer::new(value);
+        self
+    }
 
-        let full_name = self.config.path_for(&self.storage_name).file_name();
-        let shm = match SharedMemoryBuilder::new(&full_name).open_existing(AccessMode::ReadWrite) {
-            Ok(v) => v,
-            Err(SharedMemoryCreationError::DoesNotExist) => {
-                return Err(DynamicStorageOpenError::DoesNotExist);
-            }
-            Err(SharedMemoryCreationError::InsufficientPermissions) => {
-                return Err(DynamicStorageOpenError::InitializationNotYetFinalized);
-            }
-            Err(_) => {
-                fail!(from self, with DynamicStorageOpenError::Open, "{} since the underlying shared memory could not be opened.", msg);
-            }
-        };
+    fn timeout(mut self, value: Duration) -> Self {
+        self.timeout = value;
+        self
+    }
 
-        let required_size = std::mem::size_of::<Data<T>>() + self.supplementary_size;
-        if shm.size() < required_size {
-            fail!(from self, with DynamicStorageOpenError::InternalError,
-                "{} since the actual size {} does not match the required size of {}.", msg, shm.size(), required_size);
-        }
+    fn supplementary_size(mut self, value: usize) -> Self {
+        self.supplementary_size = value;
+        self
+    }
 
-        let init_state = shm.base_address().as_ptr() as *const Data<T>;
-        if unsafe { &(*init_state) }
-            .state
-            .load(std::sync::atomic::Ordering::Relaxed)
-            != IS_INITIALIZED_STATE_VALUE
-        {
-            return Err(DynamicStorageOpenError::InitializationNotYetFinalized);
-        }
+    fn create(mut self, initial_value: T) -> Result<Storage<T>, DynamicStorageCreateError> {
+        self.create_impl(initial_value)
+    }
 
-        Ok(Storage {
-            shm,
-            name: self.storage_name,
-            _phantom_data: PhantomData,
-        })
+    fn open(self) -> Result<Storage<T>, DynamicStorageOpenError> {
+        self.open_impl()
+    }
+
+    fn open_or_create(
+        mut self,
+        initial_value: T,
+    ) -> Result<Storage<T>, DynamicStorageOpenOrCreateError> {
+        match self.open_impl() {
+            Ok(storage) => Ok(storage),
+            Err(DynamicStorageOpenError::DoesNotExist) => match self.create_impl(initial_value) {
+                Ok(storage) => Ok(storage),
+                Err(DynamicStorageCreateError::AlreadyExists) => Ok(self.open_impl()?),
+                Err(e) => Err(e.into()),
+            },
+            Err(e) => Err(e.into()),
+        }
```

Comment:
If you really want to ensure to mitigate a race, this needs to run in a loop since it can always happen that there is a competing creation/deletion of the shared memory.

A comment of todo would also be fine in this case since it will be quite unlikely that something like that will actually happen in reality.